virtualizarr/manifests/store.py

```diff
@@ -0,0 +1,364 @@
+    raise ValueError(
+        f"Expected the one of stores.keys() to match the data prefix, got {stores.keys()} and {request_key}"
+    )
+
+
+class ManifestStore(Store):
+    """A Zarr store that uses obstore for fast read/write from AWS, GCP, Azure.
+
+    Parameters
+    ----------
+    stores : dict[prefix, obstore.store.ObjectStore]
+        A mapping of url prefixes to obstore Store instances set up with the proper credentials.
+
+    Warnings
+    --------
+    ManifestStore is experimental and subject to API changes without notice. Please
+    raise an issue with any comments/concerns about the store.
+
+    Notes
+    -----
+    Modified from https://github.com/zarr-developers/zarr-python/pull/1661
+    """
+
+    _manifest_group: ManifestGroup
+    _stores: StoreDict
+
+    def __eq__(self, value: object):
+        NotImplementedError
+
+    def __init__(
+        self,
+        manifest_group: ManifestGroup,
+        *,
+        stores: StoreDict,
+    ) -> None:
+        import obstore as obs
+
+        for store in stores.values():
+            if not isinstance(
+                store,
+                (
+                    obs.store.AzureStore,
+                    obs.store.GCSStore,
+                    obs.store.HTTPStore,
+                    obs.store.S3Store,
+                    obs.store.LocalStore,
+                    obs.store.MemoryStore,
+                ),
+            ):
+                raise TypeError(f"expected ObjectStore class, got {store!r}")
+        # TODO: Don't allow stores with prefix
+        # TODO: Type check the manifest arrays
+        super().__init__(read_only=True)
+        self._stores = stores
+        self._manifest_group = manifest_group
+
+    def __str__(self) -> str:
+        return f"ManifestStore({self._manifest_group}, {self._stores})"
+
+    def __getstate__(self) -> dict[Any, Any]:
+        state = self.__dict__.copy()
+        stores = state["_stores"].copy()
+        for k, v in stores.items():
+            stores[k] = pickle.dumps(v)
+        state["_stores"] = stores
+        return state
+
+    def __setstate__(self, state: dict[Any, Any]) -> None:
+        stores = state["_stores"].copy()
+        for k, v in stores.items():
+            stores[k] = pickle.loads(v)
+        state["_stores"] = stores
+        self.__dict__.update(state)
+
+    async def get(
+        self,
+        key: str,
+        prototype: BufferPrototype,
+        byte_range: ByteRequest | None = None,
+    ) -> Buffer | None:
+        # docstring inherited
+        import obstore as obs
+
+        if "zarr.json" in key:
+            return get_zarr_metadata(self._manifest_group, key)
+        var, chunk_key = parse_manifest_index(key)
+        path = self._manifest_group._manifest_dict[var]._manifest._paths[*chunk_key]
+        offset = self._manifest_group._manifest_dict[var]._manifest._offsets[*chunk_key]
+        length = self._manifest_group._manifest_dict[var]._manifest._lengths[*chunk_key]
```

Comment:
Let's split this up for readability?

```suggestion
        marr = self._manifest_group._manifest_dict[var]
        manifest = marr._manifest

        path = manifest._paths[*chunk_key]
        offset = manifest._offsets[*chunk_key]
        length = manifest._lengths[*chunk_key]
```